### ISSUE TITLE
`Yesod.Auth.Email`'s postRegisterR seems to require CSRF token when using JSON even if defaultCsrfMiddleware is not used. Is this intended or a bug?

### ISSUE THREAD

**Issue by Wizek** [yesodweb/yesod#1379]
Relevant code section:
https://github.com/yesodweb/yesod/blame/master/yesod-auth/Yesod/Auth/Email.hs#L470

```
{
    "message": "Permission Denied. A valid CSRF token wasn't present in HTTP headers or POST parameters. Because the request could have been forged, it's been rejected altogether. Check the Yesod.Core.Handler docs of the yesod-core package for details on CSRF protection."
}
```

**Comment by psibi** [comment]
It was intended as I felt using csrf check was a better default for Auth related modules. Relevant cookbook entry based on that: https://github.com/yesodweb/yesod-cookbook/blob/master/cookbook/ajax-csrf.md

Is there any reason why you consider it as a bug ?

**Comment by Wizek** [comment]
Hmm, yeah, it makes some sense to default to using CSRF for auth.

Although, if I don't have `yesodMiddleware = defaultCsrfMiddleware` enabled, is there any way I can obtain the token needed by the auth module for JSON requests? AFAIK `defaultCsrfMiddleware` is the one that puts the token among the HTTP headers. So if that middleware is not enabled then is there no way for a client to get it?

And if I do add `defaultCsrfMiddleware`, then it seems now all my non-`GET` requests suddenly require the tokens. This also makes sense, but can I configure this somehow, so that the token is only required ***&*** available for auth requests?

Or, can I disable the token check for the auth requests making `defaultCsrfMiddleware` unnecessary? I mean only in my specific case, with some configuration in my app, while leaving the global default as token required for everyone else.

Or some other idea?

And if all I write above is correct, then isn't it a pity that one can compile an application where the token is required but not available? Wouldn't it be better if it was a {compile time/init time/auth request time} {error/warning}?

**Comment by psibi** [comment]
> but can I configure this somehow, so that the token is only required & available for auth requests?

You can do something like this:

```
  yesodMiddleware app = do
    mroute <- getCurrentRoute
    let toApply =
          case mroute of
            Just (HomeR _) -> False
            _ -> True
    defaultYesodMiddleware $
      (if toApply
         then defaultCsrfMiddleware
         else id)
        app
```

**Comment by MaxGabriel** [comment]
You can use [defaultCsrfSetCookieMiddleware](https://www.stackage.org/haddock/lts-8.13/yesod-core-1.4.33/Yesod-Core.html#v:defaultCsrfSetCookieMiddleware) to set the CSRF token as a cookie, which will make it available to Javascript. This middleware will not check for a CSRF token in requests. I think this sounds like the best solution.

I'm not sure, but I think it's possible to disable all CSRF protection by implementing `makeSessionBackend _ = return Nothing`, but that would also disable sessions which it sounds like you might be using.

**Comment by Wizek** [comment]
@psibi that looks promising, I may give it a try.

@MaxGabriel Doesn't setting it on a cookie negate the usefulness of the token in the first place? E.g. the attack vector mentioned in the first minutes of this video would be exposed if it were a cookie, no? https://www.youtube.com/watch?v=hW2ONyxAySY

**Comment by MaxGabriel** [comment]
I don't think so, because the attacker can't read the value of the CSRF token in the cookie. This SO post explains it pretty well http://stackoverflow.com/a/39926988/1176156

**Comment by Wizek** [comment]
@MaxGabriel Oh, of course, so it comes down in the cookie, but must be sent up as a header, I realize now.

**Comment by MaxGabriel** [comment]
Yeah it's all pretty confusing!
